contributors/design-proposals/thirdpartyresources.md

```diff
@@ -0,0 +1,224 @@
+	Terminating ThirdPartyConditionType = "Terminating"
+)
+
+// ThirdPartyCondition contains details for the current condition of this pod.
+type ThirdPartyCondition struct {
+	// Type is the type of the condition.
+	Type ThirdPartyConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=ThirdPartyConditionType"`
+	// Status is the status of the condition.
+	// Can be True, False, Unknown.
+	Status ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status,casttype=ConditionStatus"`
+	// Last time we probed the condition.
+	// +optional
+	LastProbeTime metav1.Time `json:"lastProbeTime,omitempty" protobuf:"bytes,3,opt,name=lastProbeTime"`
+	// Last time the condition transitioned from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,4,opt,name=lastTransitionTime"`
+	// Unique, one-word, CamelCase reason for the condition's last transition.
+	// +optional
+	Reason string `json:"reason,omitempty" protobuf:"bytes,5,opt,name=reason"`
+	// Human-readable message indicating details about last transition.
+	// +optional
+	Message string `json:"message,omitempty" protobuf:"bytes,6,opt,name=message"`
+}
+
+// ThirdPartyStatus indicates the state of the ThirdParty
+type ThirdPartyStatus struct {
+	// Conditions indicate state for particular aspects of a ThirdParty
+	Conditions []ThirdPartyCondition `json:"conditions" protobuf:"bytes,1,opt,name=conditions"`
+}
+
+// +genclient=true
+
+// ThirdParty represents a resource that should be exposed on the API server.  Its name MUST be in the format
+// <.spec.name>.<.spec.group>.
+type ThirdParty struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Spec describes how the user wants the resources to appear
+	Spec ThirdPartySpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	// Status indicates the actual state of the ThirdParty
+	Status ThirdPartyStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// ThirdPartyList is a list of ThirdParty objects.
+type ThirdPartyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Items individual ThirdParties
+	Items []ThirdParty `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+```
+
+
+## Behavior
+### Create
+When a new TPR-registration is completed, no synchronous action is taken.
+A controller will run to confirm that value-space of the reserved names doesn't
+collide and sets the "NameConflict" condition to `false`.
+
+A custom `http.Handler` which accepts a delegate (like our current
+genericapiserver), will look at the `request.RequestInfo` (in the context) and
+check its Lister (created by generated client code) for a match with proper
+conditions.  If no match is found, it delegates.  If a match is found, it runs
```

Comment:
You want to say that we use the genericapiserver delegation mechanism to intercept requests by checking a TPR lister? In other words, you want to get rid of the existing `InstallAPI` handler installation for each TPR, but use a dynamic http handler filter?